### PR TITLE
ux: Improve XCM tx send flow

### PR DESCRIPTION
**Pull request by NunoAlexandre**
### Main changes

- Morph the `SendButton` to show status updates
- When sending an XCM, move the tx to 'ongoing' as soon as possible 
- Lots of (hopefully) improvements

### CODE REVIEW

File: app/src/components/Button.tsx [hunk 1]
```diff
@@ -66,7 +71,14 @@ const Button: FC<ButtonProps> = ({
   variant = 'primary',
   size = 'lg',
   children,
-  type = 'button', // Default type is 'button',
+  icon = (
+    <LoadingIcon
+      className="mr-3 animate-spin"
+      width={spinnerSize[size]}
+      height={spinnerSize[size]}
+    />
```
Comment (NunoAlexandre) [inline app/src/components/Button.tsx:79]: Worth checking if we can style the button svg color differently based on the button variant, to avoid us having to pass it explicitly with the desired color

File: app/src/components/Menu.tsx [hunk 2]
```diff
@@ -8,14 +8,14 @@ const Menu = ({
   setNewTransferInit,
   hasCompletedTransfers,
 }: DisplaysTransfers) => {
-  const initNewTransaction = newTransferInit === TransferTab.New
+  const initNewTransaction = newTransferInit === 'New'
```
Comment (NunoAlexandre) [inline app/src/components/Menu.tsx:11]: I should have done these changes in a separate PR, my bad.

Comment (NunoAlexandre) [reply app/src/components/Menu.tsx:11]: I was trying to move the ongoing txs to a new 'Ongoing' tab but I didn't like the looks of it but kept these changes moving from enum to type.

File: app/src/components/SendButton.tsx [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  const [{ label, icon, variant }, setProps] = useState<Props>(getProps(status, props.label))
+
+  useEffect(() => {
+    setProps(getProps(status, props.label))
+  }, [status, disabled])
+
+  return (
+    <div className="w-full">
+      <Button {...props} variant={variant} disabled={disabled} label={label} icon={icon} />
+    </div>
```
Comment (NunoAlexandre) [inline app/src/components/SendButton.tsx:19]: Simplify

File: app/src/hooks/useParaspellApi.tsx [hunk 4]
```diff
@@ -33,77 +34,107 @@ const useParaspellApi = () => {
...
-            })
+          try {
+            const eventsData = handleSubmittableEvents(result)
+            if (eventsData) {
+              const { messageHash, messageId, extrinsicIndex } = eventsData
 
-            // metrics
-            if (environment === Environment.Mainnet) {
-              trackTransferMetrics({
+              addOngoingTx({
```
Comment (NunoAlexandre) [inline app/src/hooks/useParaspellApi.tsx:83]: Document

File: app/src/hooks/useTransferForm.tsx [hunk 5]
```diff
@@ -190,8 +193,17 @@ const useTransferForm = () => {
         recipient: recipient,
         fees,
         onSuccess: () => {
```
Comment (NunoAlexandre) [inline app/src/hooks/useTransferForm.tsx:195]: `onSuccess` is a misleading name

File: app/src/models/transfer.ts [hunk 6]
```diff
@@ -86,10 +86,7 @@ export interface AmountInfo {
   inDollars: number
 }
 
-export enum TransferTab {
-  New = 'New',
-  Completed = 'Completed',
-}
+export type TransferTab = 'New' | 'Ongoing' | 'Done'
```
Comment (NunoAlexandre) [inline app/src/models/transfer.ts:89]:
```suggestion
export type TransferTab = 'New' | 'Done'
```

File: app/tailwind.config.ts [hunk 7]
```diff
@@ -91,7 +91,12 @@ const config: Config = {
     },
   },
   darkMode: 'class',
-  plugins: [addVariablesForColors, nextui(), require('tailwindcss-animate')],
+  plugins: [
+    addVariablesForColors,
+    nextui(),
+    require('tailwindcss-animate'),
+    require('tailwindcss-motion'),
```
Comment (NunoAlexandre) [inline app/tailwind.config.ts:98]: We should probably remove this lib, not worth the extra weight for now I'd say

File: app/src/components/SendButton.tsx [hunk 8]
```diff
@@ -0,0 +1,85 @@
+import { FC, ReactNode, useEffect, useState } from 'react'
+import { Status } from '@/hooks/useTransfer'
+import Button, { ButtonProps, ButtonVariant } from './Button'
+import LoadingIcon from './svg/LoadingIcon'
+import { colors } from '../../tailwind.config'
+
+type SendButtonProps = ButtonProps & { status: Status }
+
+const SendButton: FC<SendButtonProps> = ({ status, disabled, ...props }) => {
+  const [{ label, icon, variant }, setProps] = useState<Props>(getProps(status, props.label))
```
Comment (noahjoeris) [inline app/src/components/SendButton.tsx:10]: Hmm. Props is usually the name for Component parameters only. I have never seen it in useState. Is it possible to rename it? Also an opportunity to be more precise.

Comment (noahjoeris) [reply app/src/components/SendButton.tsx:10]: Ahh I see. You are storing the props of the other button component.

File: app/src/components/SendButton.tsx [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+  }, [status, disabled])
+
+  return <Button {...props} variant={variant} disabled={disabled} label={label} icon={icon} />
+}
+
+interface Props {
+  label?: string
+  icon?: ReactNode
+  variant?: ButtonVariant
+}
```
Comment (noahjoeris) [inline app/src/components/SendButton.tsx:23]: We can use the interface of the `Button `component. No need to recreate it here. Maybe also worth thinking about renaming the component from `Button `to `BaseButton ` or similar

File: app/src/components/SendButton.tsx [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+function getProps(status: Status, defaultLabel?: string): Props {
```
Comment (noahjoeris) [inline app/src/components/SendButton.tsx:25]:
```suggestion
function getButtonProps(status: Status, defaultLabel?: string): Props {
```

File: app/src/hooks/useParaspellApi.tsx [hunk 11]
```diff
@@ -11,10 +11,11 @@ import { txWasCancelled } from '@/utils/transfer'
 import { captureException } from '@sentry/nextjs'
 import useNotification from './useNotification'
 import useOngoingTransfers from './useOngoingTransfers'
-import { Status, TransferParams } from './useTransfer'
+import { Sender, Status, TransferParams } from './useTransfer'
+import { ISubmittableResult } from '@polkadot/types/types'
 
 const useParaspellApi = () => {
-  const { addTransfer: addTransferToStorage } = useOngoingTransfers()
+  const { addOrUpdate } = useOngoingTransfers()
```
Comment (noahjoeris) [inline app/src/hooks/useParaspellApi.tsx:18]: maybe rename?
`addOrUpdateOngoingTransfers`

File: app/src/components/SendButton.tsx (hunk 8 above)
Comment (NunoAlexandre) [reply app/src/components/SendButton.tsx:10]: I actually renamed it already to `OverwrittenProps` before this comment

File: app/src/components/SendButton.tsx (hunk 9 above)
Comment (NunoAlexandre) [reply app/src/components/SendButton.tsx:23]: > We can use the interface of the Button component

What do you suggest? We could use the `Exclude` approach but that seems more of an hassle that a benefit for example

File: app/src/components/SendButton.tsx (hunk 10 above)
Comment (NunoAlexandre) [reply app/src/components/SendButton.tsx:25]: I don't like this. This is a local function in the `SendButton` props so there's no need to add "Button" again to the naming here imo

File: app/tailwind.config.ts (hunk 7 above)
Comment (noahjoeris) [reply app/tailwind.config.ts:98]: which one? The tailwind motion? What did you use it for?

File: app/src/hooks/useParaspellApi.tsx (hunk 11 above)
Comment (NunoAlexandre) [reply app/src/hooks/useParaspellApi.tsx:18]: I did that first but I find it's unnecessarily too long. Either `addOrUpdateOngoingTx` or we leave it as is. It's a click away to understand what it is

**Review by noahjoeris — APPROVED** [review]
nice work @NunoAlexandre 
Left some small comments

File: app/src/components/SendButton.tsx (hunk 10 above)
Comment (noahjoeris) [reply app/src/components/SendButton.tsx:25]: But I was very confused to be honest. If we are talking about props in a component I am thinking of the props of THIS component. Not another one.

File: app/src/components/SendButton.tsx (hunk 8 above)
Comment (NunoAlexandre) [reply app/src/components/SendButton.tsx]: I forgot to commit & push - done now 👍

File: app/src/components/SendButton.tsx (hunk 9 above)
Comment (NunoAlexandre) [reply app/src/components/SendButton.tsx]: > Maybe also worth thinking about renaming the component from Button to BaseButton  or similar


I think that's unnecessary. `Button` is already a "base button", as the name suggests (in the sense that there is nothing specific about its name). `SendButton` is a clear subset of a `Button`. So I vote to keep it as is,

File: app/src/components/SendButton.tsx (hunk 8 above)
Comment (NunoAlexandre) [reply app/src/components/SendButton.tsx]: > Ahh I see. You are storing the props of the other button component.

Kinda. I am handling the props that make each button state unique or specific so that we can overwrite it and thus morph it.

File: app/src/components/Transfer.tsx [hunk 12]
```diff
@@ -331,19 +332,21 @@ const Transfer: FC = () => {
           tokenAmount={tokenAmount}
           fees={fees}
           durationEstimate={durationEstimate}
+          className={transferStatus === 'Idle' ? '' : 'opacity-30'}
```
Comment (Victor-Salomon) [inline app/src/components/Transfer.tsx:335]: (eventually to avoid passing the empty string)
import clsx from 'clsx';
className={clsx({ 'opacity-30': transferStatus !== 'Idle' })}

File: app/src/hooks/useParaspellApi.tsx [hunk 13]
```diff
@@ -29,81 +30,113 @@ const useParaspellApi = () => {
...
+      ? 'Transfer a̶p̶p̶r̶o̶v̶e̶d rejected'
+      : 'Failed to submit the transfer'
+
+    captureException(e)
     addNotification({
-      message: 'Failed to submit the transfer',
+      message,
       severity: NotificationSeverity.Error,
     })
   }
```
Comment (Victor-Salomon) [inline app/src/hooks/useParaspellApi.tsx:142]: heavy to review but looks promising 🤩

File: app/src/store/ongoingTransfersStore.ts [hunk 14]
```diff
@@ -48,19 +49,26 @@ export const useOngoingTransfersStore = create<State>()(
       transfers: [],
 
       // Actions
-      addTransfer: newOngoingTransfer => {
+      addOrUpdate: newOngoingTransfer => {
```
Comment (Victor-Salomon) [inline app/src/store/ongoingTransfersStore.ts:52]: Let's split this into two actions when we outsource the transfer storage from the user browser to another solution 🦾

**Review by Victor-Salomon — APPROVED** [review]
Rad one 💥
Not much to change! 
Can't wait to test it.

File: app/src/store/ongoingTransfersStore.ts (hunk 14 above)
Comment (NunoAlexandre) [reply app/src/store/ongoingTransfersStore.ts:52]: yeah I agree. I can also split it now, the major motivation being performance but given that we shouldn't have many ongoing txs at once that should not be a practical problem

Comment (Victor-Salomon) [reply app/src/store/ongoingTransfersStore.ts:52]: Considering we will need to update later, I would keep it this way for now :)

File: app/src/hooks/useParaspellApi.tsx (hunk 13 above)
Comment (NunoAlexandre) [reply app/src/hooks/useParaspellApi.tsx:154]: Yeah it's all cleaned up now, all error handling happening in one single place on this module

Comment (Victor-Salomon) [reply app/src/hooks/useParaspellApi.tsx:154]: I was speaking about the complete hook 🦾

File: app/src/components/SendButton.tsx (hunk 10 above)
Comment (NunoAlexandre) [reply app/src/components/SendButton.tsx]: I mean, it is the (overwritten) props of this component, but this component is a subset of another component. #allinterconnected

I renamed the function and added a docstring, let me know if that helps 👍

File: app/tailwind.config.ts (hunk 7 above)
Comment (NunoAlexandre) [reply app/tailwind.config.ts:98]: yeah - used it for 1/2 animations on the `SendButton`. I tried using tailwind-animate instead but couldn't make it nearly as nice but open to try that to save on another dependency

**Comment by NunoAlexandre** [comment]
> Rad one 💥 Not much to change! Can't wait to test it.

@Victor-Salomon I addressed the ticket to remove a failed ongoing tx [in this commit](https://github.com/velocitylabs-org/turtle/pull/194/commits/c3c37fec82d1e2e3dbbdc77341ab003a8b30c59f)

File: app/src/hooks/useParaspellApi.tsx [hunk 15]
```diff
@@ -29,81 +30,125 @@ const useParaspellApi = () => {
...
+  ) => {
+    setStatus('Idle')
     console.log('Transfer error:', e)
+    const message = txWasCancelled(sender, e)
+      ? 'Transfer a̶p̶p̶r̶o̶v̶e̶d rejected'
+      : 'Failed to submit the transfer'
+
+    if (txId) {
+      removeOngoing(txId)
+    }
```
Comment (Victor-Salomon) [inline app/src/hooks/useParaspellApi.tsx:147]: 🦾

**Review by Victor-Salomon — APPROVED** [review]
All good to me!!💥
